Self-Consistent-Field/RHF_symmetry.py

```diff
@@ -0,0 +1,192 @@
+print("\nSize of the ERI tensor will be %4.2f GB." % I_Size)
+
+# Estimate memory usage
+memory_footprint = I_Size * 1.5
+if I_Size > numpy_memory:
+    psi4.core.clean()
+    raise Exception("Estimated memory utilization (%4.2f GB) exceeds numpy_memory "
+                    "limit of %4.2f GB." % (memory_footprint, numpy_memory))
+
+# A matrix is returned for every irrep, even if the irrep is not present for
+# the molecule; these must be filtered out to avoid problems with NumPy arrays
+# that have a zero dimension.
+def filter_empty_irrep(coll):
+    return tuple(m for m in coll if all(m.shape))
+
+# The convention will be to have quantities in the spin orbital (SO) basis
+# ending with an underscore, each consisting of one matrix per irrep.
+S_ = filter_empty_irrep(mints.so_overlap().to_array())
+T_ = filter_empty_irrep(mints.so_kinetic().to_array())
+V_ = filter_empty_irrep(mints.so_potential().to_array())
+# The two-electron integrals are not blocked according to symmetry, so a
+# transformation between the atomic orbital (AO) and SO bases will be required.
+I_AO = np.asarray(mints.ao_eri())
+# In order to convert from the C1 AO basis to the symmetrized SO basis, a set
+# of matrices (one for each irrep with shape [nao, irrep_size]) is used to
+# transform the dense AO representation into the block-diagonal SO
+# representation. A SO-to-AO transformation matrix is obtained by transposing
+# the corresponding AO-to-SO transformation matrix.
+transformers_ = filter_empty_irrep(wfn.aotoso().to_array())
+
+assert len(S_) == len(T_) == len(V_) == len(transformers_)
+nirrep = len(S_)
+nsopi = [n for n in nsopi if n > 0]
+
+print('\nTotal time taken for integrals: %.3f seconds.' % (time.time() - t))
+t = time.time()
+
+# Build H_core: [Szabo:1996] Eqn. 3.153, pp. 141
+H_ = [T + V for (T, V) in zip(T_, V_)]
+
+# Build the orthogonalizer A = S^(-1/2) for each irrep.
+A_ = []
+for S in S_:
+    # Application of a function to a matrix requires transforming to
+    # the diagonal basis, applying the function to the diagonal form,
+    # then backtransformation to the original basis.
+    eigval, eigvec = np.linalg.eigh(S)
+    eigval_diag = np.diag(eigval ** (-1./2))
+    A = eigvec.dot(eigval_diag).dot(eigvec.T)
+    A_.append(A)
+
+def form_new_orbitals(A, F):
+    Fp = A.dot(F).dot(A)        # Eqn. 3.177
+    e, C2 = np.linalg.eigh(Fp)  # Solving Eqn. 1.178
+    C = A.dot(C2)               # Back transform, Eqn. 3.174
+    return C, e
+
+# Calculate initial core guess: [Szabo:1996] pp. 145
+C_ = []
+e_ = []
+for i in range(nirrep):
+    C, e = form_new_orbitals(A_[i], H_[i])
+    C_.append(C)
+    e_.append(e)
+
+# Occupations of each irrep are taken from the lowest eigenvalues (energies)
+# of the guess coefficients.
+energies_and_irreps = np.array(sorted(
+    (energy, irrep)
+    for (irrep, energies_irrep) in enumerate(e_)
+    for energy in energies_irrep
+))
+lowest_occupied = sorted(energies_and_irreps[:ndocc, 1].astype(int))
```

Comment:
I don't think we need a sort here. I think this might have turned out a little clearer. What do you think?

Reply:
Not sure why I had that, it's already sorted properly. It does need to be a list through.